compiler/hash-parser/src/lexer.rs

```diff
@@ -0,0 +1,618 @@
+use crate::token::TokenErrorKind;
+use crate::token::TokenKind;
+use crate::token::TokenResult;
+use crate::utils::*;
+use std::cell::Cell;
+use std::iter;
+
+/// Representing the end of stream, or the initial character that is set as 'prev' in
+/// a [Lexer] since there is no character before the start.
+const EOF_CHAR: char = '\0';
+
+pub(crate) struct Lexer<'a> {
+    /// Location of the lexer in the current stream.
+    offset: Cell<usize>,
+
+    contents: &'a str,
+
+    /// The previous character of the current stream, this is useful for keeping track
+    /// of state when tokenising compound inputs that rely on previous context.
+    prev: Cell<char>,
+}
+
+impl<'a> Lexer<'a> {
+    /// Create a new [Lexer] from the given string input.
+    pub(crate) fn new(contents: &'a str) -> Self {
+        Lexer {
+            offset: Cell::new(0),
+            contents,
+            prev: Cell::new(EOF_CHAR),
+        }
+    }
+
+    /// Returns a `Chars` iterator over the remaining characters.
+    // fn chars(&self) -> Chars<'a> {
+    //     self.contents.clone()
+    // }
+
+    /// Returns amount of already consumed symbols.
+    pub(crate) fn len_consumed(&self) -> usize {
+        self.offset.get()
+    }
+
+    /// Peeks the next symbol from the input stream without consuming it.
+    pub(crate) fn peek(&self) -> char {
+        self.nth_char(0)
+    }
+
+    /// Peeks the second symbol from the input stream without consuming it.
+    pub(crate) fn peek_second(&self) -> char {
+        self.nth_char(1)
+    }
+
+    // /// Returns nth character relative to the current position.
+    // /// If requested position doesn't exist, `EOF_CHAR` is returned.
+    // /// However, getting `EOF_CHAR` doesn't always mean actual end of file,
+    // /// it should be checked with `is_eof` method.
+    fn nth_char(&self, n: usize) -> char {
+        let offset = self.offset.get();
+
+        // @@Safety: We rely that the byte offset is correctly computed when stepping over the
+        //           characters in the iterator.
+        let slice = unsafe { std::str::from_utf8_unchecked(&self.contents.as_bytes()[offset..]) };
+
+        slice.chars().nth(n).unwrap_or(EOF_CHAR)
+    }
+
+    /// Moves to the next character.
+    pub(crate) fn next(&self) -> Option<char> {
+        let offset = self.offset.get();
+
+        // @@Safety: We rely that the byte offset is correctly computed when stepping over the
+        //           characters in the iterator.
+        let slice = unsafe { std::str::from_utf8_unchecked(&self.contents.as_bytes()[offset..]) };
+        let ch = slice.chars().next()?;
+
+        // only increment the offset by one if there is a next character
+        self.prev.set(ch);
+
+        self.offset.set(self.offset.get() + ch.len_utf8());
+        Some(ch)
+    }
+
+    /// Checks if there is nothing more to consume.
+    pub(crate) fn is_eof(&self) -> bool {
+        self.contents.len() == self.len_consumed()
```

Comment:
Will offset ever reach `self.contents.len()`? Won't it always be `<= self.contents.len() - 1`?